In [1]:
import pandas as pd
import statsmodels.api as sm

### Build a regression model

In [2]:
model_df = pd.read_csv('../data/statsmodel_df.csv')
model_df.head()

station_name  fsq_avg_rating  \
0                   de la Commune / King        8.888889   
1  de la Commune / Place Jacques-Cartier        8.833333   
2                       Berri / Cherrier        8.862500   
3              St-Dominique / St-Zotique        8.444444   
4                Duluth / de l'Esplanade        8.912500   

   fsq_avg_distance (m)  yelp_avg_rating  yelp_avg_distance (m)  total_bikes  
0            467.222222         8.611111             519.888219        11178  
1            135.333333         8.058824             232.070538         8058  
2            470.250000         8.500000             510.764503         5760  
3            147.444444         8.450000             280.756109         5760  
4            248.875000         8.250000             451.355694         5600

In [3]:
# The dependent variable is "total_bikes"
y_statsmodel =  model_df['total_bikes']

# The indepndent variables are "fsq_avg_rating", "fsq_avg_distance (m)", "yelp_avg_rating", "yelp_avg_distance (m)"
# The data quality check (null values, missing values, inconsistent types, and so on) was done previously
x_statsmodel = model_df[['fsq_avg_rating', 'fsq_avg_distance (m)', 'yelp_avg_rating', 'yelp_avg_distance (m)']]

# Add a constant (intercept) to the list of indepndent variables
x_statsmodel = sm.add_constant(x_statsmodel)
x_statsmodel.head()

const  fsq_avg_rating  fsq_avg_distance (m)  yelp_avg_rating  \
0    1.0        8.888889            467.222222         8.611111   
1    1.0        8.833333            135.333333         8.058824   
2    1.0        8.862500            470.250000         8.500000   
3    1.0        8.444444            147.444444         8.450000   
4    1.0        8.912500            248.875000         8.250000   

   yelp_avg_distance (m)  
0             519.888219  
1             232.070538  
2             510.764503  
3             280.756109  
4             451.355694

### Provide model output and interpretation of results 

In [4]:
# Create and fit the model using statsmodels linear regression model (OLS)
model = sm.OLS(y_statsmodel, x_statsmodel)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:            total_bikes   R-squared:                       0.506
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     84.85
Date:                Mon, 25 Sep 2023   Prob (F-statistic):           1.59e-49
Time:                        05:16:32   Log-Likelihood:                -2816.4
No. Observations:                 336   AIC:                             5643.
Df Residuals:                     331   BIC:                             5662.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                 -5920.74

The dependent variable "total_bikes" (also known as the target) is what we are trying to predict or explain using the independent variables (predictors).

##### R-squared
R-squared (also known as the coefficient of determination) represents the proportion of variance in the dependent variable "total_bikes" that is predictable from the inpendent variables "fsq_avg_rating", "fsq_avg_distance", "yelp_avg_rating", "yelp_avg_distance". The higher the R-squared value (range: from 0 to 1), the better our model fits the data. In other words, it is a measure of the goodness of fit of the model. For this specific model output, approximately 50.6% of the variability in the total number of bikes at a station can be explained.

##### Adjusted R-squared
Adjusted R-squared is similar to R-squared but accounts for the number of predictor variables in the model. Unlike R-squared which tends to increase with the addition of more predictors, even if they don't significantly contribute to explaining the variation in the dependent variable, adjusted R-squared penalizes the inclusion of unnecessary variables. For this specific model, the adjusted R-squared is 50% which means it can explain about half of the variability, considering the trade-off between model complexity (number of predictor variables) and goodness of fit.

##### F-statistic and Prob (F-statistic)
The F-statistic tests the overall significance of the regression model, it assess whether at least one independent variable significantly impacts the dependent variable. The higher the F-statistic the better the indication of the overall fit of the model. The Prob (F-statistic) is the probability associated with the F-statistic, a low p-value (< 0.05) indicates that the overall model is statistically significant.

##### coef
For coef's, the const (intercept) represents the value of "total_bikes" when all independent variables are 0. For each 1-meter increase in "fsq_avg_rating" and "yelp_avg_rating" we expect an approximate increase of 886.7350 and 259.3661 in the total number of bikes, respectively. This means higher ratings are associated with more bikes. For each 1-meter increase in "fsq_avg_distance (m)" and "yelp_avg_distance (m)" we expect an approximate decrease of 1.0372 and 0.5340 in the total number of bikes, respectively. This means that greater distances (from station and place) are associated with fewer bikes.

##### t and P>|t|
Statistical significance can be provided by t-statistics and associated p-values (P>|t|). Small p-values (less than 0.05 for our analysis) suggest that the predictor is statistically significant, in such a case we reject the null hypothesis. In the context of this problem, the null hypothesis (H0) states that there is no relationship or influence of that predictor on the total number of bikes. However, for all columns we can see that their p-values are less than 0.05, this means that the independent variables are statistically significant in predicting the total number of bikes (the dependent variable).

##### Skew and Kurtosis
Skew is a measure of symmetry of the distribution, in our case the skewness of 1.530 suggests that the distribution is moderately skewed to the right. Kurtosis is a measure of the sharpness of the distribution, a high positive kurtosis indicates heavy tails and a peaky distribution, in this context, the distribution contains more outliers or extreme values compared to a normal distribution.

# Stretch

### How can you turn the regression model into a classification model?

To turn the regression model into a classification we need to determine how to categorize the continous output from the regression into distinct classes.

1. Define Classes for Classification
    - Decide how to categorize the continous output, this could be based on a certain threshold. For instance, we can categorize blood alcohol content into categories: "low", "medium", and "high". 
    - Determine if it is a logistic regression model for binary classification or multinomial regression for multinomial classification.
2. Discretize the Output and Assign Labels
    - Use defined thresholds to 'bin' the continous output and assign the 'labels' a value. For example, "low": 0, "medium": 1, "high": 2.
3. Prepare the data and Conduction Feature Selection
    -  This can include determining and selecting the most relevant features for predicting the assigned class labels.
4. Choose a Classification Algorithm and Train the Model
    - Select the suitable classification algorithm for the data and problem at hand, then accordingly train the chosen classification model using the selected features and assigned labels.
5. Evaluate the Classification Model
    - Evaluate the performance of the model using appropriate metrics to ensure accurate results.
6. Adjustments and Optimization
    - Fine-tune the model, adjust parameters or different algorithms to optimize performance.
7. Make predictions
    - Use the model to make predictions on new data through processes such as interpolation.

In [5]:
# In the context of our regression model, we can turn this model into a logistic regression model for binary classification
# We can create a threshold for our dependent variable
# Convert total_bikes to binary classes (1 for total_bikes > 5000 and 0 for total_bikes < 5000)
threshold = 5000
# "model_df['total_bikes'] > threshold" returns True or False and "astype(int)" converts result to 1 or 0
y_statsmodel_labels =  (model_df['total_bikes'] > threshold).astype(int)
y_statsmodel_labels

0      1
1      1
2      1
3      1
4      1
      ..
331    0
332    0
333    0
334    0
335    0
Name: total_bikes, Length: 336, dtype: int32

In [6]:
model = sm.Logit(y_statsmodel_labels, x_statsmodel)
results = model.fit()

# Print the summary of the logistic regression model
print(results.summary())

Optimization terminated successfully.
         Current function value: 0.101060
         Iterations 12
                           Logit Regression Results                           
Dep. Variable:            total_bikes   No. Observations:                  336
Model:                          Logit   Df Residuals:                      331
Method:                           MLE   Df Model:                            4
Date:                Mon, 25 Sep 2023   Pseudo R-squ.:                  0.2453
Time:                        05:25:52   Log-Likelihood:                -33.956
converged:                       True   LL-Null:                       -44.995
Covariance Type:            nonrobust   LLR p-value:                 0.0001934
                            coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                   -36.0270     20.835     -1.729      0.084     -76.862       4

Log of odds (also known as logit) is a way of representing the odds (the likelihood of an event happening) on a logarithmic scale. Odds are the ratio of the probability of an event occurring to the probability of it not occurring. In this context, we are trying to find the probability that a bike count exceeds a certain threshold. The logistic regression model predicts the log odds of this event happening given the predictor variables.

"fsq_avg_rating" has a coef of 5.5372, for each unit increase, the log-odds of having a total number of bikes greater than 5000 increases by approximately 5.5372. In the statistical significance (p-values) for this output, it can be seen that "fsq_avg_distance (m)" and "yelp_avg_distance (m)" are not statistically significant and therefore the predictors might not be meaningful in predicting whether the total number of bikes exceeds 5000.